File: 动态规划/背包/P1833.py
# 混合背包 求最大值
import sys

# 此题在洛谷上由于Python的性能问题，需要使用C++实现才能AC

def parse_time(time):
    hour,minute = map(int,time.split(':'))
    return hour*60+minute

def main():
    start,end,n = sys.stdin.readline().split()
    T = parse_time(end)-parse_time(start)
    n = int(n)

    # dp[i] 表示不超过 i 时的最大价值
    dp = [0]*(T+1)

    for _ in range(n):
        t,val,c = map(int,sys.stdin.readline().split())

        if c == 0:
            # 完全背包
            for capacity in range(t,T+1):
                dp[capacity] = max(dp[capacity],dp[capacity-t]+val)
        else:
            # 多重背包 (多重背包使用二进制优化成为01背包)
            k = 1
            while k <= c:
                w = k*t
                value = k*val

                for capacity in range(T,w-1,-1):
                    dp[capacity] = max(dp[capacity],dp[capacity-w]+value)

                c -= k
                k *= 2

            if c > 0:
                # 处理剩余的物品
                w = c*t
                value = c*val
                
                for capacity in range(T,w-1,-1):
                    dp[capacity] = max(dp[capacity],dp[capacity-w]+value)

    sys.stdout.write(str(dp[T]))

if __name__ == '__main__':
    main()

File: 动态规划/背包/P1833.cpp
#include <bits/stdc++.h>
using namespace std;

int parse_time(const string &time)
{
    size_t pos = time.find(':');

    int hour = stoi(time.substr(0, pos));
    int minute = stoi(time.substr(pos + 1));

    return hour * 60 + minute;
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    string start, finish;
    int n;

    cin >> start >> finish >> n;

    int T = parse_time(finish) - parse_time(start);

    vector<long long> dp(T + 1, 0);

    for (int i = 0; i < n; i++)
    {
        int t;
        long long value;
        long long count;

        cin >> t >> value >> count;

        if (count == 0)
        {
            // 完全背包
            for (int capacity = t; capacity <= T; capacity++)
            {
                dp[capacity] = max(
                    dp[capacity],
                    dp[capacity - t] + value);
            }
        }
        else
        {
            // 多重背包：二进制拆分
            long long group = 1;

            while (group <= count)
            {
                long long weight = group * t;
                long long total_value = group * value;

                if (weight <= T)
                {
                    for (int capacity = T;
                         capacity >= weight;
                         capacity--)
                    {
                        dp[capacity] = max(
                            dp[capacity],
                            dp[capacity - weight] + total_value);
                    }
                }

                count -= group;
                group *= 2;
            }

            // 处理剩余物品
            if (count > 0)
            {
                long long weight = count * t;
                long long total_value = count * value;

                if (weight <= T)
                {
                    for (int capacity = T;
                         capacity >= weight;
                         capacity--)
                    {
                        dp[capacity] = max(
                            dp[capacity],
                            dp[capacity - weight] + total_value);
                    }
                }
            }
        }
    }

    cout << dp[T] << '\n';

    return 0;
}